File: lib/es6/belt_SetDict.js


import * as Belt_internalAVLset from "./belt_internalAVLset.js";

function add(t, x, cmp) {
  if (t === undefined) {
    return Belt_internalAVLset.singleton(x);
  }
  let k = t.v;
  let c = cmp(x, k);
  if (c === 0) {
    return t;
  }
  let l = t.l;
  let r = t.r;
  if (c < 0) {
    let ll = add(l, x, cmp);
    if (ll === l) {
      return t;
    } else {
      return Belt_internalAVLset.bal(ll, k, r);
    }
  }
  let rr = add(r, x, cmp);
  if (rr === r) {
    return t;
  } else {
    return Belt_internalAVLset.bal(l, k, rr);
  }
}

function remove(t, x, cmp) {
  if (t === undefined) {
    return t;
  }
  let v = t.v;
  let l = t.l;
  let r = t.r;
  let c = cmp(x, v);
  if (c === 0) {
    if (l === undefined) {
      return r;
    }
    if (r === undefined) {
      return l;
    }
    let v$1 = {
      contents: r.v
    };
    let r$1 = Belt_internalAVLset.removeMinAuxWithRef(r, v$1);
    return Belt_internalAVLset.bal(l, v$1.contents, r$1);
  }
  if (c < 0) {
    let ll = remove(l, x, cmp);
    if (ll === l) {
      return t;
    } else {
      return Belt_internalAVLset.bal(ll, v, r);
    }
  }
  let rr = remove(r, x, cmp);
  if (rr === r) {
    return t;
  } else {
    return Belt_internalAVLset.bal(l, v, rr);
  }
}

function mergeMany(h, arr, cmp) {
  let len = arr.length;
  let v = h;
  for(let i = 0; i < len; ++i){
    let key = arr[i];
    v = add(v, key, cmp);
  }
  return v;
}

function removeMany(h, arr, cmp) {
  let len = arr.length;
  let v = h;
  for(let i = 0; i < len; ++i){
    let key = arr[i];
    v = remove(v, key, cmp);
  }
  return v;
}

function splitAuxNoPivot(cmp, n, x) {
  let v = n.v;
  let l = n.l;
  let r = n.r;
  let c = cmp(x, v);
  if (c === 0) {
    return [
      l,
      r
    ];
  }
  if (c < 0) {
    if (l === undefined) {
      return [
        undefined,
        n
      ];
    }
    let match = splitAuxNoPivot(cmp, l, x);
    return [
      match[0],
      Belt_internalAVLset.joinShared(match[1], v, r)
    ];
  }
  if (r === undefined) {
    return [
      n,
      undefined
    ];
  }
  let match$1 = splitAuxNoPivot(cmp, r, x);
  return [
    Belt_internalAVLset.joinShared(l, v, match$1[0]),
    match$1[1]
  ];
}

function splitAuxPivot(cmp, n, x, pres) {
  let v = n.v;
  let l = n.l;
  let r = n.r;
  let c = cmp(x, v);
  if (c === 0) {
    pres.contents = true;
    return [
      l,
      r
    ];
  }
  if (c < 0) {
    if (l === undefined) {
      return [
        undefined,
        n
      ];
    }
    let match = splitAuxPivot(cmp, l, x, pres);
    return [
      match[0],
      Belt_internalAVLset.joinShared(match[1], v, r)
    ];
  }
  if (r === undefined) {
    return [
      n,
      undefined
    ];
  }
  let match$1 = splitAuxPivot(cmp, r, x, pres);
  return [
    Belt_internalAVLset.joinShared(l, v, match$1[0]),
    match$1[1]
  ];
}

function split(t, x, cmp) {
  if (t === undefined) {
    return [
      [
        undefined,
        undefined
      ],
      false
    ];
  }
  let pres = {
    contents: false
  };
  let v = splitAuxPivot(cmp, t, x, pres);
  return [
    v,
    pres.contents
  ];
}

function union(s1, s2, cmp) {
  if (s1 === undefined) {
    return s2;
  }
  if (s2 === undefined) {
    return s1;
  }
  let h1 = s1.h;
  let h2 = s2.h;
  if (h1 >= h2) {
    if (h2 === 1) {
      return add(s1, s2.v, cmp);
    }
    let v1 = s1.v;
    let l1 = s1.l;
    let r1 = s1.r;
    let match = splitAuxNoPivot(cmp, s2, v1);
    return Belt_internalAVLset.joinShared(union(l1, match[0], cmp), v1, union(r1, match[1], cmp));
  }
  if (h1 === 1) {
    return add(s2, s1.v, cmp);
  }
  let v2 = s2.v;
  let l2 = s2.l;
  let r2 = s2.r;
  let match$1 = splitAuxNoPivot(cmp, s1, v2);
  return Belt_internalAVLset.joinShared(union(match$1[0], l2, cmp), v2, union(match$1[1], r2, cmp));
}

function intersect(s1, s2, cmp) {
  if (s1 === undefined) {
    return;
  }
  if (s2 === undefined) {
    return;
  }
  let v1 = s1.v;
  let l1 = s1.l;
  let r1 = s1.r;
  let pres = {
    contents: false
  };
  let match = splitAuxPivot(cmp, s2, v1, pres);
  let ll = intersect(l1, match[0], cmp);
  let rr = intersect(r1, match[1], cmp);
  if (pres.contents) {
    return Belt_internalAVLset.joinShared(ll, v1, rr);
  } else {
    return Belt_internalAVLset.concatShared(ll, rr);
  }
}

function diff(s1, s2, cmp) {
  if (s1 === undefined) {
    return s1;
  }
  if (s2 === undefined) {
    return s1;
  }
  let v1 = s1.v;
  let l1 = s1.l;
  let r1 = s1.r;
  let pres = {
    contents: false
  };
  let match = splitAuxPivot(cmp, s2, v1, pres);
  let ll = diff(l1, match[0], cmp);
  let rr = diff(r1, match[1], cmp);
  if (pres.contents) {
    return Belt_internalAVLset.concatShared(ll, rr);
  } else {
    return Belt_internalAVLset.joinShared(ll, v1, rr);
  }
}

let empty;

let fromArray = Belt_internalAVLset.fromArray;

let fromSortedArrayUnsafe = Belt_internalAVLset.fromSortedArrayUnsafe;

let isEmpty = Belt_internalAVLset.isEmpty;

let has = Belt_internalAVLset.has;

let subset = Belt_internalAVLset.subset;

let cmp = Belt_internalAVLset.cmp;

let eq = Belt_internalAVLset.eq;

let forEachU = Belt_internalAVLset.forEachU;

let forEach = Belt_internalAVLset.forEach;

let reduceU = Belt_internalAVLset.reduceU;

let reduce = Belt_internalAVLset.reduce;

let everyU = Belt_internalAVLset.everyU;

let every = Belt_internalAVLset.every;

let someU = Belt_internalAVLset.someU;

let some = Belt_internalAVLset.some;

let keepU = Belt_internalAVLset.keepSharedU;

let keep = Belt_internalAVLset.keepShared;

let partitionU = Belt_internalAVLset.partitionSharedU;

let partition = Belt_internalAVLset.partitionShared;

let size = Belt_internalAVLset.size;

let toList = Belt_internalAVLset.toList;

let toArray = Belt_internalAVLset.toArray;

let minimum = Belt_internalAVLset.minimum;

let minUndefined = Belt_internalAVLset.minUndefined;

let maximum = Belt_internalAVLset.maximum;

let maxUndefined = Belt_internalAVLset.maxUndefined;

let get = Belt_internalAVLset.get;

let getUndefined = Belt_internalAVLset.getUndefined;

let getExn = Belt_internalAVLset.getExn;

let checkInvariantInternal = Belt_internalAVLset.checkInvariantInternal;

export {
  empty,
  fromArray,
  fromSortedArrayUnsafe,
  isEmpty,
  has,
  add,
  mergeMany,
  remove,
  removeMany,
  union,
  intersect,
  diff,
  subset,
  cmp,
  eq,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  keepU,
  keep,
  partitionU,
  partition,
  size,
  toList,
  toArray,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getExn,
  split,
  checkInvariantInternal,
}
/* No side effect */

File: jscomp/others/belt_SetDict.ml
(* Copyright (C) 2017 Hongbo Zhang, Authors of ReScript
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

module N = Belt_internalAVLset
module A = Belt_Array

type ('k,'id) t = 'k N.t

type ('key, 'id) cmp = ('key, 'id)  Belt_Id.cmp


(* here we relies on reference transparence
   address equality means everything equal across time
   no need to call `bal` again
*)  
let rec add  (t : _ t) x  ~cmp : _ t =
  match t with 
  | None -> N.singleton x 
  | Some nt ->
    let k = nt.value in 
    let c = (Belt_Id.getCmpInternal cmp) x k [@bs] in
    if c = 0 then t
    else
      let {N.left = l; right = r} = nt  in 
      if c < 0 then 
        let ll = add ~cmp l x in 
        if ll == l then t 
        else N.bal ll k r 
      else 
        let rr = add ~cmp r x in 
        if rr == r then t 
        else N.bal l k rr 

let rec remove (t : _ t) x  ~cmp : _ t = 
  match t with 
    None -> t
  | Some n  ->
    let {N.left = l; value = v; right = r} = n in 
    let c = (Belt_Id.getCmpInternal cmp) x v [@bs] in
    if c = 0 then 
      match l, r with 
      | (None, _) -> r 
      | (_, None) -> l 
      | (_, Some rn) -> 
        let v = ref rn.value in 
        let r = N.removeMinAuxWithRef rn v in 
        N.bal l v.contents r 
    else
    if c < 0 then 
      let ll = remove ~cmp  l x in 
      if ll == l then t
      else N.bal ll v r 
    else
      let rr = remove ~cmp  r x in 
      if rr == r then t  
      else N.bal l v rr

let mergeMany   h arr ~cmp =   
  let len = A.length arr in 
  let v = ref h in  
  for i = 0 to len - 1 do 
    let key = A.getUnsafe arr i in 
    v .contents<- add v.contents  ~cmp key 
  done ;
  v.contents 

let removeMany h arr ~cmp = 
  let len = A.length arr in 
  let v = ref h in  
  for i = 0 to len - 1 do 
    let key = A.getUnsafe arr i in 
    v .contents<- remove v.contents  ~cmp key 
  done ;
  v.contents 

let rec splitAuxNoPivot ~cmp (n : _ N.node) x : _ *  _ =   
  let {N.left = l; value = v; right = r} = n in  
  let c = (Belt_Id.getCmpInternal cmp) x v [@bs] in
  if c = 0 then l,r
  else 
  if c < 0 then
    match l with 
    | None -> 
      None ,  Some n
    | Some l -> 
      let (ll,  rl) = splitAuxNoPivot ~cmp  l x in 
      ll,  N.joinShared rl v r
  else
    match r with 
    | None ->
      Some n,  None
    | Some r -> 
      let lr,  rr = splitAuxNoPivot ~cmp  r x in 
      N.joinShared l v lr, rr

let rec splitAuxPivot ~cmp (n : _ N.node) x pres : _ *  _ =   
  let {N.left = l; value = v; right = r} = n in  
  let c = (Belt_Id.getCmpInternal cmp) x v [@bs] in
  if c = 0 then 
    begin
      pres .contents<- true;
      l, r
    end
  else 
  if c < 0 then
    match l with 
    | None -> 
      None , Some n
    | Some l -> 
      let (ll, rl) = splitAuxPivot ~cmp  l x pres in 
      ll,  N.joinShared rl v r
  else
    match r with 
    | None ->
      Some n,  None
    | Some r -> 
      let lr, rr = splitAuxPivot ~cmp  r x pres in 
      N.joinShared l v lr,  rr

let split  (t : _ t) x  ~cmp  =
  match t with 
    None ->
    (None, None), false
  | Some n ->
    let pres = ref false in 
    let v = splitAuxPivot ~cmp n x  pres in 
    v, pres.contents

(* `union s1 s2`
   Use the pivot to split the smaller collection
*)      
let rec union (s1 : _ t) (s2 : _ t) ~cmp : _ t =
  match s1, s2 with
    (None, _) -> s2
  | (_, None) -> s1
  | Some n1, Some n2 ->
    let h1, h2 = n1.height , n2.height in                 
    if h1 >= h2 then
      if h2 = 1 then add ~cmp s1 n2.value
      else begin
        let {N.left = l1; value =  v1; right = r1} = n1 in      
        let l2, r2 = splitAuxNoPivot ~cmp n2 v1 in
        N.joinShared (union ~cmp l1 l2) v1 (union ~cmp r1 r2)
      end
    else
    if h1 = 1 then add s2 ~cmp n1.value
    else begin
      let {N.left = l2; value = v2; right = r2 } = n2 in 
      let l1, r1 = splitAuxNoPivot ~cmp n1 v2  in
      N.joinShared (union ~cmp l1 l2) v2 (union ~cmp r1 r2)
    end

let rec intersect  (s1 : _ t) (s2 : _ t) ~cmp =
  match s1, s2 with
  | None, _ 
  | _, None -> None
  | Some n1, Some n2  ->
    let {N.left = l1; value = v1; right = r1 } = n1 in  
    let pres = ref false in 
    let l2,r2 = splitAuxPivot ~cmp n2 v1 pres in 
    let ll = intersect ~cmp l1 l2 in 
    let rr = intersect ~cmp r1 r2 in 
    if pres.contents then N.joinShared ll v1 rr 
    else N.concatShared ll rr 

let rec diff s1 s2 ~cmp  =
  match s1, s2 with
    (None, _) 
  | (_, None) -> s1
  | Some n1, Some n2  ->
    let {N.left = l1; value = v1; right = r1} = n1 in
    let pres = ref false in 
    let l2, r2 = splitAuxPivot ~cmp n2 v1 pres in 
    let ll = diff ~cmp l1 l2 in 
    let rr = diff ~cmp r1 r2 in 
    if pres.contents then N.concatShared ll rr 
    else N.joinShared ll v1 rr 


let empty = None     
let fromArray = N.fromArray
let isEmpty = N.isEmpty



let cmp = N.cmp
let eq = N.eq
let has = N.has
let forEachU = N.forEachU
let forEach = N.forEach
let reduceU = N.reduceU                
let reduce = N.reduce
let everyU = N.everyU               
let every = N.every
let someU = N.someU                  
let some = N.some    
let size = N.size
let toList = N.toList
let toArray = N.toArray
let minimum = N.minimum
let maximum = N.maximum
let maxUndefined = N.maxUndefined
let minUndefined = N.minUndefined
let get = N.get
let getExn = N.getExn
let getUndefined = N.getUndefined
                

let fromSortedArrayUnsafe = N.fromSortedArrayUnsafe
let subset = N.subset
let keep = N.keepShared
let keepU = N.keepSharedU             
let partitionU = N.partitionSharedU
let partition = N.partitionShared 

let checkInvariantInternal = N.checkInvariantInternal
